sympy/utilities/iterables.py

```diff
@@ -2546,8 +2546,25 @@ def necklaces(n, k, free=False):
     .. [1] http://mathworld.wolfram.com/Necklace.html
 
     """
-    return uniq(minlex(i, directed=not free) for i in
-        variations(list(range(k)), n, repetition=True))
+    # The FKM algorithm
+    if k == 0 and n > 0:
+        return
+    a = [0]*n
+    yield tuple(a)
+    if n == 0:
+        return
+    while 1:
+        i = n - 1
+        while a[i] == k - 1:
+            i -= 1
+            if i == -1:
+                return
+        a[i] += 1
+        for j in range(n - i - 1):
+            a[j + i + 1] = a[j]
+        if n % (i + 1) == 0 and (not free or all(a <= a[j::-1] + a[-1:j:-1] for j in range(n - 1))):
```

Comment:
I tested, without failure
```python
for n in range(6):
    for k in range(n+1):
        for b in range(2):
            assert all(i==j for i,j in zip(necklaces(n,k,b), old_necklaces(n,k,b)))
```
This is very nice!

I wonder if that `all(a <= ...` could be improved, perhaps, by finiding the minimal reversed string and comparing (using a better [`minlex'](#https://en.wikipedia.org/wiki/Lexicographically_minimal_string_rotation#Booth.27s_Algorithm))). But my testing only yields about a 10% improvement only at larger necklaces. The current code is straightforward and a nice improvement.

Reply:
+1 to adding reference

Reply:
Thanks. I’m not sure how to improve the bracelet test, so I’ll leave it as is. Will add the reference.